packages/assets-controllers/src/TokenRatesController.ts

```diff
@@ -421,194 +367,195 @@ export class TokenRatesController extends StaticIntervalPollingController<TokenR
-    chainIdAndNativeCurrency: {
-      chainId: Hex;
-      nativeCurrency: string;
-    }[],
-  ) {
-    await this.updateExchangeRatesByChainId(chainIdAndNativeCurrency);
-  }
-
-  /**
-   * Updates exchange rates for all tokens.
-   *
-   * @param chainIds - The chain IDs.
-   * @returns A promise that resolves when all chain updates complete.
-   */
-  /**
-   * Updates exchange rates for all tokens.
-   *
-   * @param chainIdAndNativeCurrency - The chain ID and native currency.
-   */
-  async updateExchangeRatesByChainId(
-    chainIdAndNativeCurrency: {
-      chainId: Hex;
-      nativeCurrency: string;
-    }[],
+    chainIdAndNativeCurrency: ChainIdAndNativeCurrency[],
   ): Promise<void> {
     if (this.#disabled) {
       return;
     }
 
-    // Create a promise for each chainId to fetch exchange rates.
-    const updatePromises = chainIdAndNativeCurrency.map(
-      async ({ chainId, nativeCurrency }) => {
-        const tokenAddresses = this.#getTokenAddresses(chainId);
-        // Build a unique key based on chainId, nativeCurrency, and the number of token addresses.
-        const updateKey: `${Hex}:${string}` = `${chainId}:${nativeCurrency}:${tokenAddresses.length}`;
-
-        if (updateKey in this.#inProcessExchangeRateUpdates) {
-          // Await any ongoing update to avoid redundant work.
-          await this.#inProcessExchangeRateUpdates[updateKey];
-          return null;
+    const marketData: Record<Hex, Record<Hex, MarketDataDetails>> = {};
+    const assetsByNativeCurrency: Record<
+      string,
+      {
+        chainId: Hex;
+        tokenAddress: Hex;
+      }[]
+    > = {};
+    const unsupportedAssetsByNativeCurrency: Record<
+      string,
+      {
+        chainId: Hex;
+        tokenAddress: Hex;
+      }[]
+    > = {};
+    for (const { chainId, nativeCurrency } of chainIdAndNativeCurrency) {
+      if (this.#tokenPricesService.validateChainIdSupported(chainId)) {
+        for (const tokenAddress of this.#getTokenAddresses(chainId)) {
+          if (
+            this.#tokenPricesService.validateCurrencySupported(nativeCurrency)
+          ) {
+            (assetsByNativeCurrency[nativeCurrency] ??= []).push({
+              chainId,
+              tokenAddress,
+            });
+          } else {
+            (unsupportedAssetsByNativeCurrency[nativeCurrency] ??= []).push({
+              chainId,
+              tokenAddress,
+            });
+          }
         }
+      }
+    }
 
-        // Create a deferred promise to track this update.
-        const {
-          promise: inProgressUpdate,
-          resolve: updateSucceeded,
-          reject: updateFailed,
-        } = createDeferredPromise({ suppressUnhandledRejection: true });
-        this.#inProcessExchangeRateUpdates[updateKey] = inProgressUpdate;
-
-        try {
-          const contractInformations = await this.#fetchAndMapExchangeRates({
-            tokenAddresses,
-            chainId,
+    const promises = [
```

Comment:
Whats the number of calls/promises we are generating here?
If this is large, thoughts on using `reduceInBatchesSerially` or one of the other batching mechanisms we have to process multiple promises in parallel?